src/Features/Core/Portable/LambdaSimplifier/LambdaSimplifierDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,148 @@
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterOperationAction(AnalyzeOperation, OperationKind.LambdaExpression);
+        }
+
+        private void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            var lambdaExpression = (ILambdaExpression)context.Operation;
+            var invocationExpression = TryGetInvocationExpression(lambdaExpression);
+            if (invocationExpression == null || 
+                invocationExpression.TargetMethod == null || 
+                invocationExpression.IsInvalid)
+            {
+                return;
+            }
+
+            var lambdaParameters = lambdaExpression.Signature.Parameters;
+            var invocationArguments = invocationExpression.ArgumentsInParameterOrder;
+
+            if (lambdaParameters.Length != invocationArguments.Length)
+            {
+                return;
+            }
+
+            for (int i = 0, n = lambdaParameters.Length; i < n; i++)
+            {
+                var lambdaParameter = lambdaParameters[i];
+                var argument = invocationArguments[i];
+
+                if (lambdaParameter == null || argument == null)
+                {
+                    return;
+                }
+
+                var argumentValue = UnwrapImplicitConversion(argument.Value) as IParameterReferenceExpression;
+                if (!lambdaParameter.Equals(argumentValue?.Parameter))
+                {
+                    return;
+                }
+
+                if (argument.Parameter.RefKind != RefKind.None)
+                {
+                    // Can't simplify it if one of the parameters is passed to a ref/out param.
+                    return;
+                }
+            }
+
+            var diagnostic = Diagnostic.Create(s_descriptor, lambdaExpression.Syntax.GetLocation());
+            context.ReportDiagnostic(diagnostic);
+        }
+
+        private static IOperation UnwrapImplicitConversion(IOperation value)
+        {
+            var conversion = value as IConversionExpression;
+            return conversion == null
+                ? value
+                : conversion.IsExplicit
+                    ? null
+                    : conversion.Operand;
+        }
+
+        public static IInvocationExpression TryGetInvocationExpression(ILambdaExpression lambdaExpression)
+        {
+            var body = lambdaExpression?.Body;
+            if (body?.Statements.Length != 1)
+            {
+                return null;
+            }
+
+            var firstStatement = body.Statements[0];
+            return lambdaExpression.Signature.ReturnsVoid
+                ? TryGetInvocationExpressionForSubLambda(firstStatement)
+                : TryGetInvocationExpressionForFuncLambda(firstStatement);
+        }
+
+        private static IInvocationExpression TryGetInvocationExpressionForSubLambda(
+            IOperation operation)
```

Comment:
I don't love this pair of methods, but maybe it's just that I haven't used `IOperation` enough yet.  Do we really need to Unwrap conversions everywhere?  Is that safe in these contexts, or is it changing meaning?